service/interpreter/stateExecutionCounter.go

```diff
@@ -135,25 +164,56 @@ func (e *StateExecutionCounter) MarkStateExecutionCompleted(state iwfidl.StateMo
 		case iwfidl.DISABLED:
 			return nil
 		case iwfidl.ENABLED_FOR_ALL:
-			e.decreaseStateIdCurrentlyExecutingCounts(state)
+			e.decreaseStateIdCurrentlyExecutingCounts(currentState)
+			shouldSkipUpsert := determineIfShouldSkipRefreshOnCompleted(nextStates, true)
+			if shouldSkipUpsert {
+				return nil
+			}
 		case iwfidl.ENABLED_FOR_STATES_WITH_WAIT_UNTIL:
 			fallthrough
 		default:
 			if compatibility.GetSkipWaitUntilApi(&options) {
 				return nil
 			} else {
-				e.decreaseStateIdCurrentlyExecutingCounts(state)
+				e.decreaseStateIdCurrentlyExecutingCounts(currentState)
+				shouldSkipUpsert := determineIfShouldSkipRefreshOnCompleted(nextStates, false)
+				if shouldSkipUpsert {
+					return nil
+				}
 			}
 		}
 	} else {
 		if config.GetDisableSystemSearchAttribute() {
 			return nil
 		} else {
-			e.decreaseStateIdCurrentlyExecutingCounts(state)
+			e.decreaseStateIdCurrentlyExecutingCounts(currentState)
 		}
 	}
 
-	return e.updateStateIdSearchAttribute()
+	return e.refreshIwfExecutingStateIdSearchAttribute()
+}
+
+func determineIfShouldSkipRefreshOnCompleted(nextStates []iwfidl.StateMovement, enabledForAll bool) bool {
+	var nonClosingNextStates []iwfidl.StateMovement
+	for _, s := range nextStates {
+		if _, ok := service.ValidClosingWorkflowStateId[s.GetStateId()]; !ok {
+			// s is not a ValidClosingWorkflowStateId
+			nonClosingNextStates = append(nonClosingNextStates, s)
+		}
+	}
+	if enabledForAll {
+		if len(nonClosingNextStates) > 0 {
+			return true
+		}
+	} else {
+		for _, s := range nonClosingNextStates {
+			if !s.StateOptions.GetSkipWaitUntil() {
```

Comment:
It should be `!compatibility.GetSkipWaitUntilApi(&options)` like you did in https://github.com/indeedeng/iwf/pull/453 

Btw, this double negative i hard to follow. maybe you can make it  `compatibility.HasWaitUntilApi(&options)` which will make it more clear in this statement -- when any of the next states has waitUntil, return true to skip refresh (because the next state will refresh immediately)

Reply:
Do we need `!compatibility.GetSkipWaitUntilApi(&options)`? This check is only for the newest version so I thought using `s.StateOptions.GetSkipWaitUntil()` was safe

Reply:
I changed it, but would like to understand better why not use `s.StateOptions.GetSkipWaitUntil()` in this case. Unless it's only a safety measure

Reply:
> This check is only for the newest version so I thought using s.StateOptions.GetSkipWaitUntil() was safe

Oh, you are right. It's not needed because of the versioning